File: iOS/Services/CloudSyncProfileStore.swift
import Foundation

struct CloudSyncProfileLinks: Equatable {
    let hostedBinId: String?
    let linkedBinId: String?

    var activeManifestId: String? {
        hostedBinId ?? linkedBinId
    }
}

struct CloudSyncProfileStore {
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func links(clientId: UUID?) -> CloudSyncProfileLinks {
        guard let clientId else { return CloudSyncProfileLinks(hostedBinId: nil, linkedBinId: nil) }
        migrateLegacyLinks(clientId: clientId)
        return CloudSyncProfileLinks(
            hostedBinId: readScoped(prefix: Self.hostedPrefix, clientId: clientId),
            linkedBinId: readScoped(prefix: Self.linkedPrefix, clientId: clientId)
        )
    }

    func activeManifestId(clientId: UUID?) -> String? {
        links(clientId: clientId).activeManifestId
    }

    func setHostedBinId(_ binId: String?, clientId: UUID) {
        writeScoped(prefix: Self.hostedPrefix, clientId: clientId, binId: binId)
    }

    func setLinkedBinId(_ binId: String?, clientId: UUID) {
        writeScoped(prefix: Self.linkedPrefix, clientId: clientId, binId: binId)
    }

    func clearLinks(clientId: UUID) {
        defaults.removeObject(forKey: scopedKey(prefix: Self.hostedPrefix, clientId: clientId))
        defaults.removeObject(forKey: scopedKey(prefix: Self.linkedPrefix, clientId: clientId))
    }

    private func migrateLegacyLinks(clientId: UUID) {
        migrateLegacyValue(
            legacyKey: Self.legacyHostedKey,
            prefix: Self.hostedPrefix,
            clientId: clientId
        )
        migrateLegacyValue(
            legacyKey: Self.legacyLinkedKey,
            prefix: Self.linkedPrefix,
            clientId: clientId
        )
    }

    private func migrateLegacyValue(legacyKey: String, prefix: String, clientId: UUID) {
        let key = scopedKey(prefix: prefix, clientId: clientId)
        let scopedValue = normalize(defaults.string(forKey: key))
        let legacyValue = normalize(defaults.string(forKey: legacyKey))
        if scopedValue == nil, let legacyValue {
            defaults.set(legacyValue, forKey: key)
        }
        defaults.removeObject(forKey: legacyKey)
    }

    private func readScoped(prefix: String, clientId: UUID) -> String? {
        normalize(defaults.string(forKey: scopedKey(prefix: prefix, clientId: clientId)))
    }

    private func writeScoped(prefix: String, clientId: UUID, binId: String?) {
        let key = scopedKey(prefix: prefix, clientId: clientId)
        guard let value = normalize(binId) else {
            defaults.removeObject(forKey: key)
            return
        }
        defaults.set(value, forKey: key)
    }

    private func scopedKey(prefix: String, clientId: UUID) -> String {
        prefix + clientId.uuidString.lowercased()
    }

    private func normalize(_ value: String?) -> String? {
        value?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
    }

    private static let hostedPrefix = "cloudSyncHostedBinId_client_"
    private static let linkedPrefix = "cloudSyncLinkedBinId_client_"
    private static let legacyHostedKey = "cloudSyncHostedBinId"
    private static let legacyLinkedKey = "cloudSyncLinkedBinId"
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: Android/app/src/main/java/com/example/supplementtracker/service/CloudSyncProfileStore.kt
package com.example.supplementtracker.service

import android.content.Context
import android.content.SharedPreferences
import java.util.UUID

data class CloudSyncProfileLinks(
    val hostedBinId: String?,
    val linkedBinId: String?
) {
    val activeManifestId: String?
        get() = hostedBinId ?: linkedBinId
}

internal interface CloudSyncProfileStorage {
    fun getString(key: String): String?
    fun putString(key: String, value: String)
    fun remove(key: String)
}

class CloudSyncProfileStore internal constructor(
    private val storage: CloudSyncProfileStorage
) {
    constructor(context: Context) : this(
        SharedPreferencesProfileStorage(OakPrefs.get(context))
    )

    fun links(clientId: UUID?): CloudSyncProfileLinks {
        if (clientId == null) return CloudSyncProfileLinks(null, null)
        migrateLegacyLinks(clientId)
        return CloudSyncProfileLinks(
            hostedBinId = readScoped(HOSTED_PREFIX, clientId),
            linkedBinId = readScoped(LINKED_PREFIX, clientId)
        )
    }
    fun activeManifestId(clientId: UUID?): String? = links(clientId).activeManifestId

    fun setHostedBinId(clientId: UUID, binId: String?) {
        writeScoped(HOSTED_PREFIX, clientId, binId)
    }

    fun setLinkedBinId(clientId: UUID, binId: String?) {
        writeScoped(LINKED_PREFIX, clientId, binId)
    }

    fun clearLinks(clientId: UUID) {
        storage.remove(scopedKey(HOSTED_PREFIX, clientId))
        storage.remove(scopedKey(LINKED_PREFIX, clientId))
    }

    private fun migrateLegacyLinks(clientId: UUID) {
        migrateLegacyValue(LEGACY_HOSTED_KEY, HOSTED_PREFIX, clientId)
        migrateLegacyValue(LEGACY_LINKED_KEY, LINKED_PREFIX, clientId)
    }

    private fun migrateLegacyValue(legacyKey: String, prefix: String, clientId: UUID) {
        val scopedKey = scopedKey(prefix, clientId)
        val scopedValue = normalize(storage.getString(scopedKey))
        val legacyValue = normalize(storage.getString(legacyKey))
        if (scopedValue == null && legacyValue != null) {
            storage.putString(scopedKey, legacyValue)
        }
        storage.remove(legacyKey)
    }

    private fun readScoped(prefix: String, clientId: UUID): String? {
        return normalize(storage.getString(scopedKey(prefix, clientId)))
    }
    private fun writeScoped(prefix: String, clientId: UUID, binId: String?) {
        val key = scopedKey(prefix, clientId)
        val value = normalize(binId)
        if (value == null) storage.remove(key) else storage.putString(key, value)
    }

    private fun scopedKey(prefix: String, clientId: UUID): String {
        return "$prefix${clientId.toString().lowercase()}"
    }

    private fun normalize(value: String?): String? {
        return value?.trim()?.takeIf { it.isNotEmpty() }
    }

    private companion object {
        const val HOSTED_PREFIX = "cloudSyncHostedBinId_client_"
        const val LINKED_PREFIX = "cloudSyncLinkedBinId_client_"
        const val LEGACY_HOSTED_KEY = "cloudSyncHostedBinId"
        const val LEGACY_LINKED_KEY = "cloudSyncLinkedBinId"
    }
}

private class SharedPreferencesProfileStorage(
    private val prefs: SharedPreferences
) : CloudSyncProfileStorage {
    override fun getString(key: String): String? = prefs.getString(key, null)
    override fun putString(key: String, value: String) {
        prefs.edit().putString(key, value).apply()
    }

    override fun remove(key: String) {
        prefs.edit().remove(key).apply()
    }
}
